FastSimulation/Tracking/plugins/FastTrackingMaskProducer.cc

```diff
@@ -0,0 +1,175 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/Common/interface/OwnVector.h"
+#include "DataFormats/TrackerRecHit2D/interface/SiTrackerGSRecHit2DCollection.h"
+#include "DataFormats/TrackerRecHit2D/interface/SiTrackerGSMatchedRecHit2DCollection.h"
+#include "DataFormats/TrackerRecHit2D/interface/SiTrackerGSRecHit2D.h"
+#include "DataFormats/TrackerRecHit2D/interface/SiTrackerGSMatchedRecHit2D.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "FastSimulation/Tracking/plugins/FastTrackingMaskProducer.h"
+#include <vector>
+#include <stdio.h>
+
+FastTrackingMaskProducer::FastTrackingMaskProducer(const edm::ParameterSet& conf) 
+  : oldHitMasks_exists_(false)
+  , oldHitCombinationMasks_exists_(false)
+  , overRideTrkQuals_(false)
+  , filterTracks_(false)
+{
+  // Main products                                                                                                                 
+  produces<std::vector<bool> >("hitMasks");
+  produces<std::vector<bool> >("hitCombinationMasks");
+
+  // Track collection                                                                                                                   
+  edm::InputTag trackCollectionTag = conf.getParameter<edm::InputTag>("trackCollection");
+  trackToken_ = consumes<reco::TrackCollection>(trackCollectionTag);
+
+  // old hit masks
+  oldHitMasks_exists_ = conf.exists("oldHitMasks");
+  if (oldHitMasks_exists_){
+    edm::InputTag hitMasksTag = conf.getParameter<edm::InputTag>("oldHitMasks");
+    hitMasksToken_ = consumes<std::vector<bool> >(hitMasksTag);
+  }
+
+  // old hit combination masks
+  oldHitCombinationMasks_exists_ = conf.exists("oldHitCombinationMasks");
+  if (  oldHitCombinationMasks_exists_){
+    edm::InputTag hitCombinationMasksTag = conf.getParameter<edm::InputTag>("oldHitCombinationMasks");
+    hitCombinationMasksToken_ = consumes<std::vector<bool> >(hitCombinationMasksTag);
+  }
+  
+  // read track quality from value map rather than from track itself
+  overRideTrkQuals_ = conf.exists("overrideTrkQuals");
+  if(  overRideTrkQuals_ ){
+    edm::InputTag trkQualsTag = conf.getParameter<edm::InputTag>("overrideTrkQuals");
+    if(trkQualsTag == edm::InputTag(""))
+      overRideTrkQuals_ = false;
+    else
+      trkQualsToken_ = consumes<edm::ValueMap<int> >(trkQualsTag);
+  }
+
+  // required track quality
+  trackQuality_=reco::TrackBase::undefQuality;
+  if (conf.exists("TrackQuality")){
+    filterTracks_=true;
+    std::string trackQualityStr = conf.getParameter<std::string>("TrackQuality");
+    if ( !trackQualityStr.empty() ) {
+      trackQuality_=reco::TrackBase::qualityByName(trackQualityStr);
+    }
+  }
+}
+
+void
+FastTrackingMaskProducer::produce(edm::Event& e, const edm::EventSetup& es)
+{
+  // the products
+  std::auto_ptr<std::vector<bool> > hitMasks(new std::vector<bool>());
```

Comment:
Hi Matthias

We use the vector<bool> because it's simple and fast,
and because full tracking uses something similar.

We know the size of the hitCombinationMask vector in advance, that's true.
I add fixing that to the task list of Shubhanshu and Anshul, for a later pr.

The size of the hitMask vector is not known in advance...
(unless one loops over the full rechit collection and checks all ids)

Cheers

Lukas

On Mon, Jul 13, 2015 at 7:33 PM, Matthias Komm <notifications@github.com>
wrote:

> In FastSimulation/Tracking/plugins/FastTrackingMaskProducer.cc
> <https://github.com/cms-sw/cmssw/pull/10166#discussion_r34488542>:
>
> > +  // required track quality
> > +  trackQuality_=reco::TrackBase::undefQuality;
> > +  if (conf.exists("TrackQuality")){
> > +    filterTracks_=true;
> > +    std::string trackQualityStr = conf.getParameter<std::string>("TrackQuality");
> > +    if ( !trackQualityStr.empty() ) {
> > +      trackQuality_=reco::TrackBase::qualityByName(trackQualityStr);
> > +    }
> > +  }
> > +}
> > +
> > +void
> > +FastTrackingMaskProducer::produce(edm::Event& e, const edm::EventSetup& es)
> > +{
> > +  // the products
> > +  std::auto_ptr<std::vector<bool> > hitMasks(new std::vector<bool>());
>
> @lveldere <https://github.com/lveldere> , @kripav
> <https://github.com/kripav> What about writing out a ValueMap instead?
> Also, you know the size of the vector already. Just initialize it directly
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10166/files#r34488542>.
>